File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        if not s:
            return ""
        
        max_len = 0
        longest_pal = ""

        for i in range(2 * len(s) - 1):
            if i % 2 == 0:
                left, right = i // 2, i // 2
            else:
                left, right  = i // 2, i // 2 + 1
                
            start_idx, length = self.find_longest_pal(s, left, right)
            
            if length > max_len:
                max_len = length
                longest_pal = s[start_idx: start_idx + length]
                
        return longest_pal
                
    def find_longest_pal(self, s, left, right):
        if s[left] != s[right]:
            return -1, -1
        
        while left >= 0 and right < len(s):
            if s[left] != s[right]:
                break
                
            left -= 1
            right += 1
            
        return left + 1, right - left - 1
            
                
            
                
                

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    public class Pair {
        public int start_idx;
        public int length;
        
        public Pair(int start_idx, int length) {
            this.start_idx = start_idx;
            this.length = length;
        }
    }

    public String longestPalindrome(String s) {
        if (s == null || s.length() == 0) {
            return "";
        }
        
        int max_len = 0;
        int start_idx = -1;
        for (int i = 0; i < 2 * s.length() - 1; i++) {
            int left, right;
            if (i % 2 == 0) {
                left = i / 2;
                right = i / 2;
            } else {
                left = i / 2;
                right = i / 2 + 1;
            }
            
            Pair answer = find_longest(s, left, right);
            
            if (answer.length > max_len) {
                max_len = answer.length;
                start_idx = answer.start_idx;
            }
        }
        
        return s.substring(start_idx, start_idx + max_len);
    }
    
    public Pair find_longest(String s, int left, int right) {
        if (s.charAt(left) != s.charAt(right)) {
            return new Pair(-1, -1);
        }
        
        while (left >= 0 && right < s.length()) {
            if (s.charAt(left) != s.charAt(right)) {
                break;
            }
            
            left -= 1;
            right += 1;
        }
        
        return new Pair(left + 1, right - left - 1);
    }
}